projectq/cengines/_optimize_test.py

```diff
@@ -147,3 +226,488 @@ def test_local_optimizer_identity_gates():
+    assert received_commands[1].qubits[0][0].id == qb0[0].id
+    assert received_commands[1].qubits[1][0].id == qb1[0].id
+    assert received_commands[2].qubits[0][0].id == qb2[0].id
+    assert received_commands[2].qubits[1][0].id == qb3[0].id
+    assert received_commands[3].qubits[0][0].id == qb2[0].id
+    assert received_commands[4].qubits[0][0].id == qb3[0].id
+
+def test_local_optimizer_commutable_circuit_Rz_example_1():
+    # Rzs should merge
+    local_optimizer = _optimize.LocalOptimizer(m=10)
+    backend = DummyEngine(save_commands=True)
+    eng = MainEngine(backend=backend, engine_list=[local_optimizer])
+    qb0 = eng.allocate_qubit()
+    qb1 = eng.allocate_qubit()
+    Rz(0.1) | qb0
+    H | qb0
+    CNOT | (qb1, qb0)
+    H | qb0
+    Rz(0.2) | qb0
+    eng.flush()
+    received_commands = []
+    # Remove Allocate and Deallocate gates
+    for cmd in backend.received_commands:
+        if not (isinstance(cmd.gate, FastForwardingGate) or
+                isinstance(cmd.gate, ClassicalInstructionGate)):
+            received_commands.append(cmd)
+    assert received_commands[0].gate == Rz(0.3)
+    assert len(received_commands) == 4
+
+def test_local_optimizer_commutable_circuit_Rz_example_2():
+    # Rzs shouldn't merge (Although in theory they should, this would require a new update.)
+    local_optimizer = _optimize.LocalOptimizer(m=10)
+    backend = DummyEngine(save_commands=True)
+    eng = MainEngine(backend=backend, engine_list=[local_optimizer])
+    qb0 = eng.allocate_qubit()
+    qb1 = eng.allocate_qubit()
+    Rz(0.1) | qb1
+    H | qb0
+    CNOT | (qb1, qb0)
+    H | qb0
+    Rz(0.2) | qb1
+    eng.flush()
+    received_commands = []
+    # Remove Allocate and Deallocate gates
+    for cmd in backend.received_commands:
+        if not (isinstance(cmd.gate, FastForwardingGate) or
+                isinstance(cmd.gate, ClassicalInstructionGate)):
+            received_commands.append(cmd)
+    assert received_commands[1].gate == Rz(0.1)
+    assert len(received_commands) == 5
+
+def test_local_optimizer_commutable_circuit_Rz_example_3():
+    # Rzs should not merge because they are operating on different qubits
+    local_optimizer = _optimize.LocalOptimizer(m=10)
+    backend = DummyEngine(save_commands=True)
+    eng = MainEngine(backend=backend, engine_list=[local_optimizer])
+    qb0 = eng.allocate_qubit()
+    qb1 = eng.allocate_qubit()
+    Rz(0.1) | qb1
+    H | qb0
+    CNOT | (qb1, qb0)
+    H | qb0
+    Rz(0.2) | qb0
+    eng.flush()
+    received_commands = []
+    # Remove Allocate and Deallocate gates
+    for cmd in backend.received_commands:
+        if not (isinstance(cmd.gate, FastForwardingGate) or
+                isinstance(cmd.gate, ClassicalInstructionGate)):
+            received_commands.append(cmd)
+    assert received_commands[1].gate == Rz(0.1)
+    assert len(received_commands) == 5
+
+def test_local_optimizer_commutable_circuit_Rz_example_4():
```

Comment:
Same comment as above about making this test parameterised.

Reply:
Done. As will be seen in next commit.